### PR TITLE
HPCC-18493 Refactor record translation code so dafilesrv can use it

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Smoketest/OBT will cover required testing

### CODE REVIEW

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+{
+    for (;;)
+    {
+        char c = *in++;
+        if (!c)
+            throw MakeStringException(0, "Invalid filter - missing closing '");
+        if (c=='\'')
+            break;
+        if (c=='\\')
+            UNIMPLEMENTED;
```
Comment (ghalliday) [inline rtl/eclrtl/eclhelper_dyn.cpp]: Should be implemented (if only to read the next char and check it isn't null)

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        : fileName(_fileName), in(_in), out(_out), chooseN(_chooseN), skipN(_skipN), rowLimit(_rowLimit)
+    {
+        inrec = &in->queryRecordAccessor(true);
+        numOffsets = inrec->getNumVarFields() + 1;
+        translator.setown(createRecordTranslator(queryOutputMeta()->queryRecordAccessor(true), *inrec));
+    }
+    virtual bool needTransform() override
+    {
+        return true;
+        //return translator->needsTranslate(); might be more appropriate?
```
Comment (ghalliday) [inline rtl/eclrtl/eclhelper_dyn.cpp]: should this be different?

File: rtl/eclrtl/eclhelper_dyn.cpp (hunk 1 above)
Comment (richardkchapman) [reply rtl/eclrtl/eclhelper_dyn.cpp]: It should - this bit is still a work in progress

File: rtl/eclrtl/eclhelper_dyn.cpp (hunk 2 above)
Comment (richardkchapman) [reply rtl/eclrtl/eclhelper_dyn.cpp]: True will work fine for now - there may be a point in the future where we could avoid it if code is refactored elsewhere, but that should be the subject of a different Jira I think.

File: rtl/eclrtl/rtlrecord.cpp [hunk 3]
```diff
@@ -495,14 +495,43 @@ COutputMetaData::~COutputMetaData()
...
     const RtlRecord *useAccessor = aRecordAccessor.load(std::memory_order_relaxed);
     if (!useAccessor)
         useAccessor = setupRecordAccessor(*this, expand, aRecordAccessor);
     return *useAccessor;
 }
 
+size32_t COutputMetaData::getRecordSize(const void * data)
+{
+    //Allocate a temporary offset array on the stack to avoid runtime overhead.
+    const RtlRecord &r = queryRecordAccessor(true);
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.cpp]: You had some code to optimize this for fixed size fields.

**Pull request merged**